shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -0,0 +1,1912 @@
+      string getMadRepresentation() {
+        exists(SummaryComponent head, SummaryComponentStack tail |
+          head = this.head() and
+          tail = this.tail() and
+          result = tail.getMadRepresentation() + "." + head.getMadRepresentation()
+        )
+        or
+        exists(SummaryComponent c |
+          this = TSingletonSummaryComponentStack(c) and
+          result = c.getMadRepresentation()
+        )
+      }
+
+      /** Gets a textual representation of this stack. */
+      string toString() { result = this.getMadRepresentation() }
+    }
+
+    /** Provides predicates for constructing stacks of summary components. */
+    module SummaryComponentStack {
+      /** Gets a singleton stack containing `c`. */
+      SummaryComponentStack singleton(SummaryComponent c) {
+        result = TSingletonSummaryComponentStack(c)
+      }
+
+      /**
+       * Gets the stack obtained by pushing `head` onto `tail`.
+       *
+       * Make sure to override `RequiredSummaryComponentStack::required()` in order
+       * to ensure that the constructed stack exists.
+       */
+      SummaryComponentStack push(SummaryComponent head, SummaryComponentStack tail) {
+        result = TConsSummaryComponentStack(head, tail)
+      }
+
+      /** Gets a singleton stack for an argument at position `pos`. */
+      SummaryComponentStack argument(ParameterPosition pos) {
+        result = singleton(SummaryComponent::argument(pos))
+      }
+
+      /** Gets a singleton stack representing a return of kind `rk`. */
+      SummaryComponentStack return(ReturnKind rk) {
+        result = singleton(SummaryComponent::return(rk))
+      }
+    }
+
+    /**
+     * A class that exists for QL technical reasons only (the IPA type used
+     * to represent component stacks needs to be bounded).
+     */
+    class RequiredSummaryComponentStack extends Unit {
+      /**
+       * Holds if the stack obtained by pushing `head` onto `tail` is required.
+       */
+      abstract predicate required(SummaryComponent head, SummaryComponentStack tail);
+    }
+
+    /**
+     * A callable with a flow summary.
+     *
+     * This interface is not meant to be used directly, instead use the public
+     * `SummarizedCallable` interface. However, _if_ you need to use this, make
+     * sure that that all classes `C` that extend `SummarizedCallableImpl` also
+     * extend `SummarizedCallable`, using the following adapter pattern:
+     *
+     * ```ql
+     * private class CAdapter extends SummarizedCallable instanceof C {
+     *   override predicate propagatesFlow(string input, string output, boolean preservesValue) {
+     *     none()
+     *   }
+     *
+     *   override predicate hasProvenance(Provenance provenance) {
+     *     C.super.hasProvenance(provenance)
+     *   }
+     * }
+     * ```
+     */
+    abstract class SummarizedCallableImpl extends SummarizedCallableBaseFinal {
```

Comment:
What does this class achieve? It's instantiation in `SummarizedCallableExternal` seems somewhat nonsensical since the `propagatesFlow` columns and `provenance` columns are formed by cartesian product. Aka, I don't think the generated vs manual logic in `SummarizedCallableExternal` does anything? That's all been moved to the language-specific side, right?

Reply:
Ah, right, I see that it handles the translation between input/output-specs of type string vs. `SummaryComponentStack`, so I'll rephrase this as a simplification comment in `SummarizedCallableExternal`.

Reply:
Indeed.